server/resp/src/main/java/org/infinispan/server/resp/operation/SetOperation.java

```diff
@@ -0,0 +1,212 @@
+package org.infinispan.server.resp.operation;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.server.resp.response.SetResponse;
+
+public class SetOperation {
+   private static final CompletableFuture<Long> NO_EXPIRATION = CompletableFuture.completedFuture(-1L);
+
+   public static CompletionStage<SetResponse> performOperation(AdvancedCache<byte[], byte[]> cache, List<byte[]> arguments) {
+      if (arguments.size() < 2) throw new IllegalArgumentException("SET operation does not have enough arguments");
+
+      SetOperationOptions options = new SetOperationOptions(arguments);
+
+      CompletionStage<Long> expirationTime;
+      // We delay the retrieval for later when operation has `XX`.
+      if (options.isKeepingTtl() && options.typeOfOperation() != SetOperationType.PUT_IF_PRESENT) {
+         expirationTime = cache.getCacheEntryAsync(options.key)
```

Comment:
This can be changed so that the thenCompose below is just a method and we only create the lambda referencing that method in this if block. The others ones can just directly invoke the method. Also then we put the thenApply code here in the same lambda, saving another allocation.

Reply:
I've updated this, although I don't think I was able to reduce the number of lambdas :(

Reply:
My original idea was instead of this SetOperation class that the state would be held in just local variables in a method and then it would invoke the correct things and we could check if the previous get or set were already completed and not allocate the lambdas. But the more I think about it I guess I am fine this way as these commands will be less used than just a regular SET most likely.